Text/CustomEntitySearch/CustomEntitySearch.cs

```diff
@@ -52,48 +53,273 @@ public static class CustomEntitySearch
+                                 int leniency = (exactMatches != null && exactMatches.Contains(word)) ? 0 : offset;
+                                 AddValues(word, text, entities, entitiesFound, leniency, caseSensitive);
+                                 if (synonyms != null && synonyms.ContainsKey(word))
+                                 {
+                                     foreach (string synonym in synonyms[word])
+                                     {
+                                         leniency = (exactMatches != null && exactMatches.Contains(synonym)) ? 0 : offset;
+                                         AddValues(synonym, text, entities, entitiesFound, leniency, caseSensitive);
+                                     }
+                                 }
+                             }
+                         }
+
+                         outRecord.Data["Entities"] = entities;
+                         outRecord.Data["EntitiesFound"] = entitiesFound;
+                     }
+                    return outRecord;
+                });
+
+            return new OkObjectResult(response);
+        }
+
+        public static void AddValues(string checkMatch, string text, List<Entity> entities, HashSet<string> entitiesFound, int leniency, bool caseSensitive)
+        {
+            bool addWord = false;
+            string escapedWord = Regex.Escape(checkMatch);
+            string pattern = (caseSensitive) ? @"\b(?x:" + escapedWord + @")\b" : @"\b(?ix:" + escapedWord + @")\b";
+            MatchCollection entityMatch = Regex.Matches(text, pattern, RegexOptions.Compiled, TimeSpan.FromSeconds(MaxRegexEvalTime));
+            if (entityMatch.Count != 0)
+            {
+                foreach (Match match in entityMatch)
+                {
+                    entities.Add(
+                        new Entity
+                        {
+                            Category = "customEntity",
+                            Value = match.Value,
+                            Offset = match.Index,
+                            Confidence = 1
+                        });
+                }
+                addWord = true;
+            }
+
+            if (leniency > 0)
+            {
+                // using KMP for now
+                // Create Table
+                IList<int> KMPTable = (caseSensitive) ? CreateKMPTable(checkMatch) : CreateKMPTable(checkMatch.ToLower());
+
+                // Begin searching!
+                int currTextChar = 0;
+                int currWordChar = 0;
+                StringWriter wordFound = new StringWriter();
+                double currMismatch = 0;
+                IList<char> wordCharArray = (caseSensitive) ? checkMatch.ToCharArray() : checkMatch.ToLower().ToCharArray();
+                IList<char> textCharArray = (caseSensitive) ? text.ToCharArray() : text.ToLower().ToCharArray();
+                IList<string> valuesList = new List<string>();
```

Comment:
nit: you should be able to do this outside the for each word loop just to save a little redundant work

Reply:
the issue with checking case sensitivity outside of the for each word loop is that I cannot populate entities with the word found in the original text

Reply:
You should always be able to use a case-insentitive version of whatever API you use to do string or chard comparisons. In principle, you should not be using `ToLower` to achieve case insensitivity.

Reply:
Sorry, I'm not as familiar. Would the non-ToLower comparison repeat code but just ignoring case? Just confused

Reply:
What I'm saying is that all string comparison APIs in .NET have an overload that allows for case insensitivity, dictionaries can be configured to be case-insensitive, etc. `ToLower` should only really be used when you're going to actually use the lowercase version. If you make sure you use those APIs, then you don't need to modify the strings and make them lower, and it should make it safe to move the code out of the loop.